### ISSUE TITLE
concurrent creation use and destruction of map/cache leads to hang

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#13873]
this test was normal passing but now are regression looks like 
concurrent creation use and destruction of objects leads to hang.

https://hazelcast-l337.ci.cloudbees.com/view/stable/job/stable-x3/81/console

/disk1/jenkins/workspace/stable-x3/3.11-SNAPSHOT/2018_10_01-12_37_04/create-use-destroy Failed

http://54.82.84.143/~jenkins/workspace/stable-x3/3.11-SNAPSHOT/2018_10_01-12_37_04/create-use-destroy

all member and client are hanging in calling create or destroy of map / cache.

**Comment by Danny-Hazelcast** [comment]
running on 3.10.5
https://hazelcast-l337.ci.cloudbees.com/view/stable/job/stable-x3/82/console
PASSED

**Comment by vbekiaris** [comment]
Can we get a run with diagnostics & slow operation stack traces turned on?

**Comment by Danny-Hazelcast** [comment]
this one failed with the same hanging https://hazelcast-l337.ci.cloudbees.com/view/stable/job/stable-x3/88/console


Oct 03, 2018 12:03:44 PM com.hazelcast.spi.ProxyService
WARNING: [10.0.0.29]:5701 [HZ] [3.11-SNAPSHOT] Error while destroying a proxy.
java.util.concurrent.TimeoutException
	at com.hazelcast.util.FutureUtil.executeWithDeadline(FutureUtil.java:376)
	at com.hazelcast.util.FutureUtil.waitWithDeadline(FutureUtil.java:363)
	at com.hazelcast.util.FutureUtil.waitWithDeadline(FutureUtil.java:337)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.destroyDistributedObject(ProxyServiceImpl.java:171)
	at com.hazelcast.spi.AbstractDistributedObject.destroy(AbstractDistributedObject.java:62)
	at hzcmd.queue.CreateUseDestroy.timeStep(CreateUseDestroy.java:24)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)

Oct 03, 2018 12:03:45 PM com.hazelcast.map.impl.EnterpriseMapServiceContextImpl
WARNING: [10.0.0.29]:5701 [HZ] [3.11-SNAPSHOT] Map partition was not destroyed in expected time, possible leak
Oct 03, 2018 12:03:45 PM com.hazelcast.map.impl.EnterpriseMapServiceContextImpl
WARNING: [10.0.0.29]:5701 [HZ] [3.11-SNAPSHOT] Map partition was not destroyed in expected time, possible leak

Oct 03, 2018 12:12:14 PM com.hazelcast.map.impl.EnterpriseMapServiceContextImpl
WARNING: [10.0.0.29]:5701 [HZ] [3.11-SNAPSHOT] Map partition was not destroyed in expected time, possible leak
Oct 03, 2018 12:12:14 PM com.hazelcast.map.impl.EnterpriseMapServiceContextImpl
WARNING: [10.0.0.29]:5701 [HZ] [3.11-SNAPSHOT] Map partition was not destroyed in expected time, possible leak

**Comment by vbekiaris** [comment]
In 3 out of 4 members' logs, apart from the expected slow map destroy operation, there is also this slow operation detected:

```
WARNING: [10.0.0.29]:5701 [HZ] [3.11-SNAPSHOT] Slow operation detected: com.hazelcast.cache.impl.operation.CachePutOperation
java.lang.Object.wait(Native Method)
        com.hazelcast.spi.impl.AbstractCompletableFuture.get(AbstractCompletableFuture.java:229)
        com.hazelcast.spi.impl.AbstractCompletableFuture.get(AbstractCompletableFuture.java:191)
        com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl.invokeOnAllMembers(EventServiceImpl.java:321)
        com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl.registerListenerInternal(EventServiceImpl.java:291)
        com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl.registerListener(EventServiceImpl.java:252)
        com.hazelcast.internal.partition.impl.PartitionEventManager.addPartitionLostListener(PartitionEventManager.java:100)
        com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl.addPartitionLostListener(InternalPartitionServiceImpl.java:1079)
        com.hazelcast.internal.partition.PartitionServiceProxy.addPartitionLostListener(PartitionServiceProxy.java:109)
        com.hazelcast.internal.eviction.ExpirationManager.scheduleExpirationTask(ExpirationManager.java:80)
        com.hazelcast.cache.impl.AbstractCacheRecordStore.markExpirable(AbstractCacheRecordStore.java:1599)
        com.hazelcast.cache.impl.CacheRecordStore.createRecord(CacheRecordStore.java:106)
        com.hazelcast.cache.impl.AbstractCacheRecordStore.createRecord(AbstractCacheRecordStore.java:653)
        com.hazelcast.cache.impl.AbstractCacheRecordStore.createRecordWithExpiry(AbstractCacheRecordStore.java:687)
        com.hazelcast.cache.impl.AbstractCacheRecordStore.createRecordWithExpiry(AbstractCacheRecordStore.java:716)
        com.hazelcast.cache.impl.AbstractCacheRecordStore.put(AbstractCacheRecordStore.java:1251)
        com.hazelcast.cache.impl.AbstractCacheRecordStore.put(AbstractCacheRecordStore.java:1282)
        com.hazelcast.cache.impl.operation.CachePutOperation.run(CachePutOperation.java:61)
        com.hazelcast.spi.Operation.call(Operation.java:170)
        com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:208)
        com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:197)
        com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:413)
        com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:153)
        com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:123)
        com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110)
```

The `partitionLostListener` registration fans out operations to all members from a partition thread (executing `CachePutOperation`) and apparently map destroy operations are blocked waiting for `MapPartitionDestroyOperation`s to complete (also executing on partition operation threads) -> deadlock that results in timeout.

**Comment by vbekiaris** [comment]
As a workaround, the partition lost listener registration can be initiated on first execution of the scheduled clear expired task, so the partition thread is not blocked (see https://github.com/vbekiaris/hazelcast/commit/2a3a46abff92fe8746eae9b0aba184397a0514d6). Using this branch, the create-use-destroy test passes (https://hazelcast-l337.ci.cloudbees.com/view/stable/job/stable-x3/93). Blocking during map destroy remains an issue.

**Comment by mdogan** [comment]
We can also forbid partition threads from calling `EventServiceImpl.registerListener`. So, we can detect similar issues earlier.

**Comment by mmedenjak** [comment]
Thank you @vbekiaris and @ahmetmircik for the fix! I would say that being this close to the release we are best to leave the map and cache destroy issue until the next release. It is definitely an issue that we should address but in the most common cases, destroy isn't called that often and structures should be long-lived. The other reason is that I suspect solving the problem might be very tricky and introduce lots of changes with unforeseen consequences.

**Comment by ahmetmircik** [comment]
@mdogan testing your suggestion here: https://github.com/hazelcast/hazelcast/pull/13898

**Comment by ahmetmircik** [comment]
@mmedenjak My PR is not sufficient to fix the main issue related to map#destroy, it only addresses a lately introduced similar issue. So Danny's test should still fail with the merged PR.

**Comment by mmedenjak** [comment]
Yes, I know that but based on [this comment](https://github.com/hazelcast/hazelcast/pull/13897#issuecomment-426898500) I'm expecting the test to pass. Also, the map destroy issue is not a new regression and I [would like to postpone that until 3.12 if possible](https://github.com/hazelcast/hazelcast/issues/13873#issuecomment-426903356).

**Comment by Danny-Hazelcast** [comment]
https://hazelcast-l337.ci.cloudbees.com/view/stable/job/stable-x3/99/console

still Failing using

Hazelcast Enterprise 3.11-SNAPSHOT (20181005 - 5a9605a, 6673239) starting

**Comment by Danny-Hazelcast** [comment]
looks like with out map and cache, only  executor and semaphore are hanging

https://hazelcast-l337.ci.cloudbees.com/view/stable/job/stable-x3/104/console

```
14:49:22 /disk1/jenkins/workspace/stable-x3/3.11-SNAPSHOT/2018_10_05-14_32_21/create-use-destroy Failed
14:49:22 
14:49:22 http://54.82.84.143/~jenkins/workspace/stable-x3/3.11-SNAPSHOT/2018_10_05-14_32_21/create-use-destroy
```

ops

http://54.82.84.143/~jenkins/workspace/stable-x3/3.11-SNAPSHOT/2018_10_05-14_32_21/create-use-destroy/createUseDestroy

**Comment by mmedenjak** [comment]
```
SEVERE: [10.0.0.11]:5701 [HZ] [3.11-SNAPSHOT] Failed to execute java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@280467e3
java.util.concurrent.RejectedExecutionException
        at com.hazelcast.util.executor.CachedExecutorServiceDelegate.execute(CachedExecutorServiceDelegate.java:111)
        at com.hazelcast.util.executor.CachedExecutorServiceDelegate.submit(CachedExecutorServiceDelegate.java:122)
        at com.hazelcast.spi.impl.executionservice.impl.DelegatingCallableTaskDecorator.call(DelegatingCallableTaskDecorator.java:49)
        at com.hazelcast.spi.impl.executionservice.impl.DelegatingCallableTaskDecorator.call(DelegatingCallableTaskDecorator.java:31)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
        at java.lang.Thread.run(Thread.java:748)
        at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
        at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)

Oct 05, 2018 11:34:20 AM com.hazelcast.scheduledexecutor.impl.ScheduledExecutorContainer
WARNING: [10.0.0.11]:5701 [HZ] [3.11-SNAPSHOT] [Scheduler: scheduledExecutorCreateUseDestroy_create-use-destroy0][Partition: 29][Task: 6d6f7e1d-29bc-4c11-9a0f-fc8de04b2c73] Error while destroying
java.util.concurrent.ExecutionException: java.util.concurrent.RejectedExecutionException
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.util.concurrent.FutureTask.get(FutureTask.java:188)
        at com.hazelcast.scheduledexecutor.impl.DelegatingScheduledFutureStripper.peel(DelegatingScheduledFutureStripper.java:95)
        at com.hazelcast.scheduledexecutor.impl.DelegatingScheduledFutureStripper.cancel(DelegatingScheduledFutureStripper.java:61)
        at com.hazelcast.scheduledexecutor.impl.ScheduledTaskDescriptor.cancel(ScheduledTaskDescriptor.java:142)
        at com.hazelcast.scheduledexecutor.impl.ScheduledExecutorContainer.destroy(ScheduledExecutorContainer.java:138)
        at com.hazelcast.scheduledexecutor.impl.AbstractScheduledExecutorContainerHolder.destroyContainer(AbstractScheduledExecutorContainerHolder.java:73)
        at com.hazelcast.scheduledexecutor.impl.DistributedScheduledExecutorService.resetPartitionOrMemberBinContainer(DistributedScheduledExecutorService.java:251)
        at com.hazelcast.scheduledexecutor.impl.DistributedScheduledExecutorService.destroyDistributedObject(DistributedScheduledExecutorService.java:183)
        at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.destroyLocalDistributedObject(ProxyServiceImpl.java:181)
        at com.hazelcast.spi.impl.proxyservice.impl.operations.DistributedObjectDestroyOperation.run(DistributedObjectDestroyOperation.java:45)
        at com.hazelcast.spi.Operation.call(Operation.java:170)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:208)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:197)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:413)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:153)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:123)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110)
Caused by: java.util.concurrent.RejectedExecutionException
        at com.hazelcast.util.executor.CachedExecutorServiceDelegate.execute(CachedExecutorServiceDelegate.java:111)
        at com.hazelcast.util.executor.CachedExecutorServiceDelegate.submit(CachedExecutorServiceDelegate.java:122)
        at com.hazelcast.spi.impl.executionservice.impl.DelegatingCallableTaskDecorator.call(DelegatingCallableTaskDecorator.java:49)
        at com.hazelcast.spi.impl.executionservice.impl.DelegatingCallableTaskDecorator.call(DelegatingCallableTaskDecorator.java:31)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
        at java.lang.Thread.run(Thread.java:748)
        at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
        at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```
@tkountis can you check it out?

**Comment by mmedenjak** [comment]
As for the semaphore:
```
"pool-6-thread-1" prio=10 tid=0x00007f26ac008800 nid=0xd4f waiting on condition [0x00007f26f5fde000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:315)
        at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:160)
        at com.hazelcast.concurrent.semaphore.SemaphoreProxy.acquire(SemaphoreProxy.java:79)
        at com.hazelcast.concurrent.semaphore.SemaphoreProxy.acquire(SemaphoreProxy.java:68)
        at hzcmd.semaphore.CreateUseDestroy.timeStep(CreateUseDestroy.java:14)
        at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
        at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
        at remote.bench.BenchThread.call(BenchThread.java:41)
        at remote.bench.BenchThread.call(BenchThread.java:14)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by Danny-Hazelcast** [comment]
@mmedenjak 
you explanation for semaphore case is correct, i have changed the test to just create destroy and it
is not longer hanging. 

@tkountis 
for the executor issue, it is not hanging.  the false report was caused by the logging


```
WARNING: [10.0.0.91]:5701 [HZ] [3.11-SNAPSHOT] [Scheduler: scheduledExecutorCreateUseDestroy_create-use-destroy0][Partition: 110][Task: 0fcf98f4-72bc-4ad8-a203-1988b1f2f27c] Error while destroying
java.util.concurrent.ExecutionException: java.util.concurrent.RejectedExecutionException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:188)
	at com.hazelcast.scheduledexecutor.impl.DelegatingScheduledFutureStripper.peel(DelegatingScheduledFutureStripper.java:95)
	at com.hazelcast.scheduledexecutor.impl.DelegatingScheduledFutureStripper.cancel(DelegatingScheduledFutureStripper.java:61)
	at com.hazelcast.scheduledexecutor.impl.ScheduledTaskDescriptor.cancel(ScheduledTaskDescriptor.java:142)
	at com.hazelcast.scheduledexecutor.impl.ScheduledExecutorContainer.destroy(ScheduledExecutorContainer.java:138)
	at com.hazelcast.scheduledexecutor.impl.AbstractScheduledExecutorContainerHolder.destroyContainer(AbstractScheduledExecutorContainerHolder.java:73)
	at com.hazelcast.scheduledexecutor.impl.DistributedScheduledExecutorService.resetPartitionOrMemberBinContainer(DistributedScheduledExecutorService.java:251)
	at com.hazelcast.scheduledexecutor.impl.DistributedScheduledExecutorService.destroyDistributedObject(DistributedScheduledExecutorService.java:183)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.destroyLocalDistributedObject(ProxyServiceImpl.java:181)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.destroyDistributedObject(ProxyServiceImpl.java:169)
	at com.hazelcast.spi.AbstractDistributedObject.destroy(AbstractDistributedObject.java:62)
	at hzcmd.executor.scheduled.CreateUseDestroy.timeStep(CreateUseDestroy.java:28)
```
as in the event of the semaphore hang. 
i search for 'at hzcmd'  which matches the above.


so,

semaphore hang test fix.

executor hang, false report.

**Comment by Danny-Hazelcast** [comment]
this issue looks to be fixed / not reproducible any more so happy to close
